source/translated_table_model.ts

```diff
@@ -157,13 +157,15 @@ export class TranslatedTableModel extends TableModel {
         return rowIndex;
       }
     })();
+    const trans = this.translation.slice();
+    const reverse = this.reverseTranslation.slice();
     for(let index = start; index < this.translation.length; ++index) {
```

Comment:
This loops from ```start``` to ```this.translation.length```, what justification is there that this is the correct range to iterate over?

Reply:
When something is inserted/removed everything above the index doesn't change. Everything below the index needs to get shifted up or down.

Reply:
How do you know that's true?

Reply:
Well when a new item is inserted room needs to be made for it. To make room for it things need to be shifted down. Shifting up doesn't make sense because then the array won't start from zero.

The reason I don't just do a splice for that is because I need to preserve the order in reverseTranslation.

Reply:
And the way I determined how to update reverseTranslation was by adding a row to translation. Then I would determine what reverseTranslation should look like with the translated row and a really obvious pattern emerged.

Reply:
Okay I see, the tricky part then is to get rid of the copies.

Reply:
Or to update the values in a more efficient way. That's a big part of what I'm struggling with.